plugins/text-editor-resources/src/components/extension/mermaid.ts

```diff
@@ -0,0 +1,680 @@
+  const pendingFragments = new Set(prev.pendingFragments)
+  const theme: MermaidConfig['theme'] = isThemeDark(getCurrentTheme()) ? 'dark' : 'default'
+
+  const unusedRenderCache = new Set(renderCache.keys())
+  const usedFragments = new Set<HTMLElement>()
+
+  const newRenderCacheEntry = getTxMeta(tr)?.renderResult
+  if (newRenderCacheEntry !== undefined) {
+    renderCache.set(newRenderCacheEntry.input, newRenderCacheEntry)
+  }
+
+  const buildOrReuseFragment = (renderState: MermaidRenderResult | null): HTMLElement | null => {
+    if (renderState?.svg === undefined) return null
+
+    const reuse = renderState.domFragments.find(f => !usedFragments.has(f))
+    if (reuse !== undefined) {
+      usedFragments.add(reuse)
+      return reuse
+    }
+
+    const container = document.createElement('div')
+    container.className = 'mermaidPreview'
+    container.innerHTML = renderState.svg
+
+    renderState.domFragments.push(container)
+    usedFragments.add(container)
+
+    return container
+  }
+
+  const buildDiagram = (node: ProseMirrorNode, nodeid: string | number, view: EditorView, theme: MermaidConfig['theme']): MermaidRenderResult => {
+    const textContent = node.textContent
+    const nodeRenderState = renderCache.get(textContent)
+    const nodeTargetRenderState =
+      nodeRenderState ??
+      prev.nodeidCache.get(nodeid) ??
+      { input: textContent, theme, domFragments: [] }
+
+    unusedRenderCache.delete(textContent)
+
+    nodeidCache.set(nodeid, nodeTargetRenderState)
+    const element = buildOrReuseFragment(nodeTargetRenderState)
+
+    if ((nodeRenderState === undefined && !pendingFragments.has(textContent)) || nodeRenderState?.theme !== theme) {
+      pendingFragments.add(textContent)
+
+      prev.throttle.call(nodeid, () => {
+        renderMermaidDiagram(textContent, theme).then(renderResult => {
+          view.dispatch(setTxMeta(view.state.tr, { renderResult }))
+          pendingFragments.delete(textContent)
+        }, (reject) => {
+          console.log(reject)
+          pendingFragments.delete(textContent)
+        })
+      })
+    }
+
+    return { ...nodeTargetRenderState, domFragments: element !== null ? [element] : [] }
+  }
+
+  const decorations: Decoration[] = []
+  const lastDecorationSet = tr !== undefined
+    ? prev.decorationSet.map(tr.mapping, tr.doc)
+    : prev.decorationSet
+
+  const nodeStatePatch = getTxMeta(tr)?.nodePatch
+
+  let mIndex = 0
+  doc.descendants((node, pos, parent, index) => {
+    if (node.type.name !== MermaidExtension.name) {
+      return
+    }
+
+    // See https://github.com/yjs/y-prosemirror/issues/49
+    // Ideally, one would hope that decorations would be recovered through the use of
+    // transaction mapping, but with Yjs this is not possible in many cases.
+    // So there's a need for dirty hacks to keep the state of the node intact. (See below).
+    const yjsdoc = options.ydoc?.getXmlFragment('content')
```

Comment:
Is that still true and reproducible? We use decorations in several places but never noticed behavior like this, but this is probably because we recreate all decorations in this case instead of mapping them.

In either case, we will need some generic solution for this to use decorations properly, and ideally I would like to avoid use of ydoc here.